### DISCUSSION TITLE
Is there a more automated way of detecting packages and dependencies?

### DISCUSSION THREAD

**Discussion by mmJordanO** [vercel/turborepo#408] [Help] [answered]
Hi.  Our team wants to migrate some apps to a monorepo solution and thus far we've got something that looks like this in the folder structure

```
apps/landing-pages
packages/config
packages/frontend
```

Effectively, `LP` depends on both packages and `frontend` also depends on the `config` package.  I was able to get something work although I think my solution was not the greatest.

In the root package.json:
```
"scripts": {
    "build": "turbo run build",
    "build:config": "cd packages/config && yarn build",
    "build:frontend": "cd packages/frontend && yarn build",
    "build:landing-pages": "turbo run build --scope=landing-pages",
    "dev": "turbo run dev --parallel",
    "dev:landing-pages": "turbo run dev --scope=landing-pages",
    "deploy": "yarn prepare && turbo run build --scope=landing-pages --includeDependencies --no-deps",
    "lint": "turbo run lint",
    "format": "prettier --write \"**/*.{ts,tsx,md}\"",
    "prepare": "yarn build:config && yarn build:frontend"
  },
```

Effectively, when someone does a fresh `yarn install`, it needs to build the `config` and then the `frontend` (in that order).  It also needs to the do the same for deployment.  

Is there a cleaner way to tell the packages what depends on what?    For instance, can I tell frontend that it needs config or even better automatically detect that during build it knows that an external package is needed so it will build said package first before proceeding?

Thanks in advance.

**Comment by fivethreeo** [comment, marked as answer]
Dependencies in package.json of each package/app decides what is dependent on what. 

So, add "config": "\*" to dependencies or devdependencies in lp and frontend add "frontend": "\*" to deps or devdeps in lp

The have "build": "dependsOn": "^build" in turbo pipelines.

Remove the toplevel build\:* scripts and keep just the build script

or add —scope=packagename to the build:\* scripts

**Reply by mmJordanO** [reply]
Thanks for the reply.  Yeah, since yesterday I talked to another person and we set up a similar thing to what you mentioned above.  I guess the pipeline part was throwing me off.

Effectively, the pipeline now has
```
"deploy": {
        "dependsOn": [
          "^build"
        ]
      },
      "dev": {
        "cache": false,
        "dependsOn": [
          "^build"
        ]
      }
```

And the individual apps and packages now have the appropriate package names in their devDependencies with the asterisks.  It builds in the appropriate order now.
